In [82]:
# Import Libraries for Neural Networks
import keras
from keras.models import Sequential
from keras.layers import Dense

# Import simple data processing libraries
import pandas as pd
import numpy as np

In [83]:
# Download the concrete dataset from the api

concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [84]:
# Define the structure of the Deep Neural Network

def regression_model(n_cols):
    # create model
    model = Sequential()

    # Input Layer
    model.add(Dense(n_cols, activation='relu', input_shape=(n_cols,)))

    # Hidden Layer
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [85]:
# Decide the predictors and Target
X = concrete_data.drop('Strength', axis=1)
y = concrete_data['Strength']

# Normalize the predictors
X_scaled = (X - X.mean()) / X.std()
X_scaled.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

## **Part A**

In [86]:
# Extract the number of features
n_cols=X.shape[1]

# Initlize the Model and fit the data on model with Validation Test set of 30% and epoches of 50
model = regression_model(n_cols)
history = model.fit(X, y, validation_split=0.3, epochs = 50, verbose=0)
MSEs = history.history['val_loss']

In [87]:
# Display the Mean and STd of Mean Squared Errors

MSEs = np.array(MSEs)
print("Mean of The Loss: Mean Squared Error is", MSEs.mean())
print("Standard Deviation of The Loss: Mean Squared Error is", MSEs.std())

Mean of The Loss: Mean Squared Error is 525.9966400146484
Standard Deviation of The Loss: Mean Squared Error is 792.3283548002006


## **Part B**

In [88]:
model2 = regression_model(n_cols)
history2 = model2.fit(X_scaled, y, validation_split=0.3, epochs=50, verbose=0)
MSEs2 = history2.history['val_loss']

In [89]:
MSEs2 = np.array(MSEs2)
print("Mean of The Loss: Mean Squared Error is", MSEs2.mean())
print("Standard Deviation of The Loss: Mean Squared Error is", MSEs2.std())

Mean of The Loss: Mean Squared Error is 455.02526000976565
Standard Deviation of The Loss: Mean Squared Error is 384.1759373569486


There is no significant Change in THe Mean, but the standard Deviation of the losses where the Part B std reduced to ~384 from ~792.



## **Part C**

In [90]:
model3 = regression_model(n_cols)
history3 = model3.fit(X_scaled, y, validation_split=0.3, epochs=100, verbose=0)
MSEs3 = history3.history['val_loss']

In [91]:
MSEs3 = np.array(MSEs3)
print("Mean of The Loss: Mean Squared Error is", MSEs3.mean())
print("Standard Deviation of The Loss: Mean Squared Error is", MSEs3.std())

Mean of The Loss: Mean Squared Error is 286.90995727539064
Standard Deviation of The Loss: Mean Squared Error is 314.43151898968966


There is highly siginicant change in MSE after increasing epoches, with difference of around ~169 in reduction of mean and ~70 in reduction of the Standard Deviation of the distribution of the means.

## **Part D**

In [100]:
def regression_model_D(n_cols):
    # create model
    model = Sequential()

    #Input Layer
    model.add(Dense(n_cols, activation='relu', input_shape=(n_cols,)))

    #Three Hidden Layers
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [101]:
# Repeating Part B but with Three hidden layers as now.

model_D = regression_model_D(n_cols)
history_4 = model_D.fit(X_scaled, y, validation_split=0.3, epochs = 50, verbose=0)
MSEs_4 = history_4.history['val_loss']

In [102]:
MSEs4 = np.array(MSEs_4)
print("Mean of The Loss: Mean Squared Error is", MSEs4.mean())
print("Standard Deviation of The Loss: Mean Squared Error is", MSEs4.std())

Mean of The Loss: Mean Squared Error is 288.6862902832031
Standard Deviation of The Loss: Mean Squared Error is 275.79411565879747


The Mean of The Mean Squared Error distribution although increased to ~2, but the Standard Deviation is indeed decreased to ~314 from ~384, suggesting that the residual errors are now more in control.